docs/style-guide.md

```diff
@@ -0,0 +1,423 @@
+
+Thus statements like:
+
+```
+let x = self.data[n];
+```
+
+are always prone to "panic" if the index/key is incorrect. (⚠️ and this includes accessing maps ⚠️).
+
+This is because these APIs return a value (or value reference) which cannot be "null", so there is no way for them to signal failure via the return type.
+
+## Non-Panicking APIs
+
+The alternative to using direct data accessors which can panic is to use a method which can return **Option** or **Result**. In the case of collections and strings, where a simple data item is being requested, this is most often provided by functions such as "get" (or "get_mut" for mutable references) which return an Option.
+
+If data access is expected to fail occasionally (e.g. looking up properties in a map) then the resulting [Option can be unwrapped](https://doc.rust-lang.org/std/option/enum.Option.html#method.unwrap_or) or propagated accordingly.
+
+If missing data signals a "hard" error from which the function cannot recover (e.g. user supplies incorrect input) then any returned Option should be [propagated into a Result immediately](https://doc.rust-lang.org/std/option/enum.Option.html#method.ok_or), with an appropriate error message.
+
+## Best Practice
+
+### Don't Panic :: required
+
+Call non-panicking data access APIs whenever data is not guaranteed to be safe.
+
+This should not include the contract of code in a different Crate. I.e. if a function in a different Crate promises to return a valid map key, but it's not a compile time checked type (like an enum), then the calling code must allow for it to fail.
+
+### Don't Handle Errors :: required
+
+Functions which can error for any reason must return a Result, and APIs should be designed such that you should never need to recover from an [Err](https://doc.rust-lang.org/std/result/enum.Result.html#variant.Err) internally (which should always be immediately propagated up to the user by using the [`?` operator](https://doc.rust-lang.org/edition-guide/rust-2018/error-handling-and-panics/the-question-mark-operator-for-easier-error-handling.html)). I.e. Never write library code which recovers from its own "errors", since if it can be recovered from, then it wasn't an "error".
+
+This approach should mean that error handling and the design of functions which can propagate errors is consistent everywhere. For non-error cases, where different types of result are possible, use a normal enum.
+
+The only time you might need to handle an **Err** is if you call APIs outside the library which return Result rather than Option (e.g. allowing a retry for data loading).
+
+Finally, and fairly obviously, **never turn an error into a panic by unconditionally unwrapping the result**.
```

Comment:
I'm not sure about this, there are cases where the API has an error state that _should never happen_ based on your inputs/invariants, and if it does, it's a bug. Unwrapping isn't intrinsically bad.

(Though it is better to use `expect()` over `unwrap()`)

Reply:
Let's aim for this and see if it stands up to real code :) It's fine to relax later after we get real examples.

Reply:
I've been thinking about this. If you can always do: obj.foo()? to propagate an Err up, I don't see why we'd even want to unwrap, even during development. Unwrapping hides the fact that the function you're in can return Err, and means you end up risking declaring an API like:
foo() -> T
rather than:
foo() -> IcuResult<T>
And then whoever's using that (maybe also in the library) thinks they don't need to care about an error, and so on. I think that being strict about "never unwrap" is good because it will force the API to visibly reflect that there can be failure.